### ISSUE TITLE
HIMSS and AHA data merge follow ups

### ISSUE THREAD

**Issue by mgarduchicago** [hospital-executives/hospital-ceos-code#5]
**Background**
- In the HIMSS build code, we also pull in an AHA dataset and join the two. 
- HIMSS does not have the AHA number, only the Medicare number. 
- First, we use a Medicare Number to AHA number crosswalk, to add in AHA numbers to the HIMSS dataset. 
- Then, we join the HIMSS data to the AHA data, on AHA number. 
- Lastly, for hospitals that fail the above criteria, we do a fuzzy match on hospital name and address, and a strict match on zip, to pull in additional matches to the AHA data. 

**For this task**

- How many hospitals from HIMSS don't match the AHA data initially?
- How many don't match after implementing the fuzzy match?
- How many hospitals in AHA don't match to HIMSS?

- [X] Add a flag for whether a hospital directly matched or fuzzy matched
- [X] export the "main" xwalk between HIMSS and AHA before any sample restrictions are made, for just the hospitals -> xwalk is located in `hospital_ceos/_data/derived/himss_aha_xwalk.feather`

**Comment by kpapen** [comment]
The algorithm to assign AHA numbers can largely be broken into two parts: generously assigning AHA numbers (mcr_aha_xwalk.py) and refining AHA numbers so that each AHA number is assigned to a single entity (himss_to_aha.R), e.g., if the UCSD hospital campus has both UCSD Medical Center and the Shiley Eye Institute (which is on the same campus) in HIMSS but only the UCSD Medical Center has an AHA number, the initial pass would assign the AHA number to both entities and the secondary pass would assign the final AHA number only to the medical center. 

1. [mcr_aha_walk.py] Assign AHA numbers if: 
    a. A CCN exists in HIMSS and for a given (CCN, year) tuple has a unique AHA match in hospital_ownership.dta
    b. A CCN exists in HIMSS and the HIMSS address is close to the AHA address
    c. TO DO - (simultaneously with fuzzy flag)
2. [himss_to_aha.R] Backfills and forward fills AHA numbers by entity_uniqueid.
3. [himss_to_aha.R] For HIMSS observations that were assigned an AHA number in Parts 1 and 2, I first rename the previously assigned AHA to be campus_aha. Then, within a campus AHA, I assign an entity its true AHA number if: 
    a. There is only one case where HIMSS entity name matches the AHA entity name and the haentitytypeid is 1. All other entity_uniqueids within the campus AHA are assigned no final AHA number.
    b. There is only one case where the HIMSS city matches the AHA city and the haentitytypeid is 1. All other entity_uniqueids within the campus AHA are assigned no final AHA number.
    c. There is only one case where the HIMSS haentitytypeid is 1. All other entity_uniqueids within the campus AHA are assigned no final AHA number. 
    d. The HIMSS entity name is most similar to the AHA entity name and the Jaro-Winkler similarity is at least 0.8
In order to ensure consistency, the clean AHA number is then backfilled according to entity_uniqueid and entity_name. Since different entities can have the same name, I restrict the names case only to hospitals with haentitytypeid == 1. The steps in Part 3 assign clean AHA numbers to 98% of the ever assigned AHA numbers from Part 1.

**Comment by mshi311** [comment]
@kpapen to make sure I understand:
The final dataset after running `himss_to_aha.R` will have both a field called `campus_aha` as well as something like `entity_aha`. In your UCSD example, the dataset would look like the following (for a given year):

| `entity_name`    | `haentitytypeid` | `campus_aha` | `entity_aha` |
| -------- | ------- | ------- | ------- |
| UCSD Medical Center  | 1   | 12345 | 12345 |
| Shiley Eye Institute | (not 1) |  12345    |  -- |

**Comment by kpapen** [comment]
Yes. Also, in the case where no `haentitytypeids` for `campus_aha` 12345 are 1, Shiley would not be assigned `entity_aha` since it's Jaro Winkler similarity score with the AHA hospital name will be less than 0.8.

**Comment by kpapen** [comment]
### Questions 1 and 2

<img width="650" height="494" alt="Image" src="https://github.com/user-attachments/assets/9cb2121a-414f-49d9-9e30-0b749599c30b" />

There are 5844 unique entities after subsetting for `haentitytypeid == 1` (i.e., hospitals). This means that in the initial match (note that this includes mapping between direct matches on characteristics such as name and zip code), 1218 don't match. After the fuzzy match, 784 don't match.

### Question 3
There are initially 7,101 hospitals in the AHA that do not appear in our subset of HIMSS. However, many of these appear to be outside of the United States; in the build code, we subset the data to exclude non-US hospitals. Roughly subsetting by coordinates, this number is reduced to 6,962.

**Comment by kpapen** [comment]
Looks good! `entity_aha` is currently `clean_aha` in our data but I will change this. I also need to double check that `entity_aha` is missing for all non-hospitals.

**Comment by kpapen** [comment]
To Do:

- [ ] Determine why there is an `entity_uniqueid` mismatch between the entity and individual level data sets. 

- [ ] Fix clean_aha and campus_aha distinctions. 
- [ ] Merge individual data on campus level.
- [ ] Create flag for the "real" hospital.

- [ ] Clean up variable names.
- [ ] Clean coordinate variables.

**Comment by kpapen** [comment]
I believe this accounts for part of @julia-paris's missing observations problem. I believe the other part is due to a lag in file updating on my end, which should also be fixed. We'll meet tomorrow to confirm this and/or determine which other entities are missing.

**Comment by kpapen** [comment]
I've been working on cleaning up the merge but it looks like the rules that lead to the most accurate matches result in a 90% vs 97% match rate from HIMSS hospitals to AHA hospitals at the (hospital, year) level. I think part of this is being driven by HIMSS reporting more hospitals (with `haentitytypeid == 1`) per AHA "hospital". I've put some examples below, but should I spend time identifying these cases so we can say it has a 90% match rate but accounting for this gives us an x% match rate? I think it will take me a bit of time to figure out so I want to check first.

Ahanumber 6140011 is reported as "hallmark health system" in the AHA at 585 Lebanon St. We assign this AHA to Melrose Wakefield Hospital since it is also at 585 Lebanon St. However, this means that Lawrence Memorial Hospital, which is part of the same system as Melrose Wakefield, is not assigned an AHA number.

6840040 is reported as "boulder community hospital" and is matched to boulder community hospital by address since the addresses match for 2014 earlier. As a result, boulder community foothills hospital, which matches the listed address for 2015 and later is not assigned an entity_aha.

6740292 is Kindred Corpus Christi and is matched by address to "Kindred Hospital Corpus Christi", but as a result, "Christus Spohn Hospital Corpus Christi South" is not assigned an entity_aha.
